### PR TITLE
Stop sending the nick through data channels after some time

### PR DESCRIPTION
**Pull request by danxuliu**
Requires #4181

~~(Kept as draft until #4181 is merged, but ready to review nevertheless)~~

Since the introduction of the HPB the nick has been sent over and over again every second, as it is not possible to know when other participants have established a data channel connection (and neither to associate a signaling id with a Nextcloud id to get the nick from the participant list).
    
However, it is possible to know when other participants join the call, and it is safe to assume that the other participant established a connection with the local one after a "reasonable" amount of time.
    
As sending the nick every second is too aggressive (specially in calls with a lot of users) now the nick is sent when the local participant establishes a connection with the remote one and, as that does not guarantee that the remote one has established a connection with the local one, the nick is sent again several times (with an increasing interval) in the next 30 seconds. If a connection is established with another participant in the meantime the cycle starts again.

Additionally a `nickChanged` signaling message* is now sent along with the data channel message. The signaling message is needed for those cases in which there is no Peer object for a participant, as in that case it is not possible to send data channel messages. However, the data channel messages still need to be sent to keep compatibility with the mobile apps.

*An unknown signaling message is just ignored by the mobile apps, but it should not break anything; tested in iOS by @ivansss, although just my guess from checking the Android app code (but I have not actually tested it).

### CODE REVIEW

**Comment by nickvergessen** [comment]
@Ivansss please check if this works with apps? (I guess they would need adjustments?)

**Review by Ivansss — APPROVED** [review]
Tested and works 👍

**Comment by nickvergessen** [comment]
Status? shall we merge it in 21 only so we test it a bit more?

**Comment by danxuliu** [comment]
> Status?

@PVince81 You self requested a review, should we wait for your review or should we merge? :-)

> shall we merge it in 21 only so we test it a bit more?

I would backport it to 20 already, but I am also fine with waiting until the next maintenance release.

**Comment by PVince81** [comment]
@danxuliu I've somehow deferred this as I wasn't sure whether you had other works in progress that would impact this PR, since you said you're going to revamp the "data channel" vs "signaling" vs ... stuff

**Comment by danxuliu** [comment]
Ah, no, no, the revamp and fixes were going to be done after this was merged (or before it was merged but on top of this anyway, not by ammending anything here). Sorry for the misunderstanding.

File: src/utils/webrtc/models/CallParticipantModel.js
Comment (PVince81) [inline src/utils/webrtc/models/CallParticipantModel.js:177]: obsolete ?

File: src/utils/webrtc/webrtc.js
Comment (PVince81) [inline src/utils/webrtc/webrtc.js:976]: is this something that we could put into a separate function to make it clearer ? or is this very specific to the nick use case ? (could be something for later)

File: src/utils/webrtc/webrtc.js
Comment (PVince81) [inline src/utils/webrtc/webrtc.js:962]: add comment to explain this

do we have different event arg formats ?

**Review by PVince81 — APPROVED** [review]
:+1: code looks fine with my limited knowledge

tested with and without signaling server with a guest whose nick changes, it seemed to work fine

I've added some optional comments

File: src/utils/webrtc/models/CallParticipantModel.js
Comment (danxuliu) [reply src/utils/webrtc/models/CallParticipantModel.js:177]: Yes, see the commit description for details: https://github.com/nextcloud/spreed/commit/e00f6627754b298edaffa135d9a20b3f8ff35e56

File: src/utils/webrtc/webrtc.js
Comment (danxuliu) [reply src/utils/webrtc/webrtc.js:976]: Yes, I will address that when I fix the signaling messages for participants without Peer objects. So something for later :-)

File: src/utils/webrtc/webrtc.js
Comment (danxuliu) [reply src/utils/webrtc/webrtc.js:962]: Yes, `nickChanged` for guests only include the name as a string, while for users it is an object with the name as a string and the user ID as a string.

I thought that the code was self-explaining, but I can still add a comment if you want.

**Comment by danxuliu** [comment]
@PVince81 Comment added; if you are satisfied with it feel free to merge, otherwise please push fixup commits to improve it ;-)

**Pull request merged**
